## Question 2 Naive Bayes Classifier

####Part 1 [CM4]

In [ ]:
# Data 1 Naive Bayes
var_smoothing = [1e-10, 1e-9, 1e-5, 1e-3, 1e-1]
nb_accuracy_seeds = []

for value in var_smoothing:
    nb_clf = GaussianNB(var_smoothing=value)
    score = cross_val_score(nb_clf, X_seeds_train, y_seeds_train, cv=10)
    nb_accuracy_seeds.append(np.mean(score))
    print(np.mean(score))

0.905
0.905
0.905
0.9175000000000001
0.8983333333333334


From the results of seeds dataset above, we can see that the highest validation accuracy (0.9175) is given by var_smoothing=1e-3. We then evaluate this on test set.

In [ ]:
nb_clf = GaussianNB(var_smoothing=1e-3)
nb_clf.fit(X_seeds_train,y_seeds_train)
print("Test accuracy of Naive Bayes for seeds dataset: ",nb_clf.score(X_seeds_test,y_seeds_test))

Test accuracy of Naive Bayes for seeds dataset:  0.925


In [ ]:
# Data 2 Naive Bayes
var_smoothing = [1e-10, 1e-9, 1e-5, 1e-3, 1e-1]
nb_accuracy_covid = []

for value in var_smoothing:
    nb_clf = GaussianNB(var_smoothing=value)
    score = cross_val_score(nb_clf, X_covid_train, y_covid_train, cv=10)
    nb_accuracy_covid.append(np.mean(score))
    print(np.mean(score))

0.4193573266629289
0.4470896766866491
0.5562277713086355
0.6038837854544932
0.6506952171475395


From the results of Covid dataset above, we can see that the highest validation accuracy (0.6507) is given by var_smoothing=1e-1. We then evaluate this on test set.

In [ ]:
nb_clf_co = GaussianNB(var_smoothing=1e-1)
nb_clf_co.fit(X_covid_train,y_covid_train)
print("Test accuracy of Naive Bayes for Covid dataset: ",nb_clf_co.score(X_covid_test,y_covid_test))

Test accuracy of Naive Bayes for Covid dataset:  0.6632756866734486


`Var_smoothing` lets our users control the portion of the largest variance of all features which is added to variances for calculation stability. Since in Gaussian Naive Bayes classifier, we assume that features are sampled from a gaussian distribution.But if the features are actually not following a normal distribution, then we can add this smoothing to kind of 'flatten' the curves and accounts for more samples that are not close to the distribution mean.

From the results of seeds dataset, we can see adding `var_smoothing` does not change the results dramatically, the accuracy does not change untill increasing `var_smoothing` to 1e-3, improving from 0.905 to 0.918, but soon decrease to 0.898 when we increase `var_smoothing` to 1e-1. This could implies that the distributions of the features of seed dataset are close to normal distribution, thus an appropriate smoothing in variance would help, but not in a daramtical way and too much smoothing would have a negative impact.

From the results of Covid dataset, we can find the accuracy improves a lot with the increase of `var smoothing`. This is mainly because some features of Covid dataset are one hot encoding categorical features which do not follow a normal distribution, increasing `var_smoothing` would help smoothing the distribution curves thus improves the accuracy. With increasing of `var_smoothing`, the accuracy of 10-fold cross validation increases from 0.419 to 0.651.

####Part 2

**For seeds dataset**

In [ ]:
nb_clf.sigma_

array([[1.47875621, 0.33480059, 0.00900017, 0.06201897, 0.04070806,
        1.68445565, 0.07707704],
       [1.92666615, 0.34365953, 0.0089558 , 0.06643817, 0.04222245,
        1.45014661, 0.05757769],
       [0.48677014, 0.11978259, 0.0091581 , 0.02713382, 0.0283718 ,
        1.60318485, 0.03533927]])

In [ ]:
nb_clf.theta_

array([[14.28269231, 14.26884615,  0.88000577,  5.49938462,  3.24076923,
         2.75613846,  5.08213462],
       [18.47509091, 16.20672727,  0.88255818,  6.18145455,  3.68961818,
         3.66296364,  6.04552727],
       [11.92980392, 13.26666667,  0.85106078,  5.23572549,  2.86598039,
         4.59947059,  5.11607843]])

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
y_pred_seed = nb_clf.predict(X_seeds_test)
conf_mat_seed = confusion_matrix(
        y_seeds_test, y_pred_seed, labels = [1, 2, 3])
conf_mat_seed

array([[13,  0,  0],
       [ 2, 11,  0],
       [ 1,  0, 13]])

From the variance results, we can find for kama(1) and Rosa(2), feature#1 area and #6 asymmetry are significantly larger than other features, while for Canadian(3), feature#6 are significantly larger than other features.

From the mean results, we can find for Kama, Rosa and Canadian, feature#1 area and #2 perimeter are much larger than the other features.

From the confusion matrix, we can see that all three classes are bascially correctly classified, with only few wrong predictions.

Thus, we can conclude that all three classes are well classifed, the important features are area, perimeter and asymmetry. Recall the conclusion we made from decision tree, We can find that they are consistent.

**For Covid dataset**

In [ ]:
nb_clf_co.sigma_

array([[4.85982857, 1.20746539, 3.13294847, 0.6855686 , 0.93530502,
        0.93530341, 0.93146352, 0.82072899, 0.84947666, 0.83000319,
        0.70230925, 0.68607672, 0.72760836, 0.69861179, 0.69213392,
        0.73130084, 0.6855686 , 0.68886219, 0.69138084, 0.69263532,
        0.6863306 , 0.71585592, 0.70573395, 0.68759801, 0.6898712 ,
        0.68658434, 0.69613065, 0.71322699, 0.73107103, 0.71680802,
        0.76535735, 0.68936695, 0.6878511 , 0.6863306 , 0.68810407,
        0.68936695, 0.71514049, 0.68658434, 0.89873729, 0.85184448,
        0.69238468, 0.70378009, 0.68759801, 0.68810407, 0.6855686 ,
        0.6878511 , 0.84977445, 0.84977445],
       [5.16526152, 1.33566681, 3.30456936, 0.68531434, 0.93525994,
        0.93525994, 0.90103474, 0.91459357, 0.80100663, 0.83666211,
        0.69488523, 0.68531434, 0.7105824 , 0.70937514, 0.68911464,
        0.70961685, 0.68683795, 0.69037493, 0.68886219, 0.69188302,
        0.6855686 , 0.71965189, 0.71250729, 0.68734478, 0.69188302,
   

In [ ]:
nb_clf_co.theta_

array([[ 3.72711089e+00,  4.37276721e+01, -7.95856928e+01,
         2.54323499e-04,  5.03051882e-01,  4.96693795e-01,
         4.37945066e-01,  1.61495422e-01,  2.07019329e-01,
         1.75483215e-01,  1.72939980e-02,  7.62970498e-04,
         4.42522889e-02,  1.34791455e-02,  6.86673449e-03,
         4.83214649e-02,  2.54323499e-04,  3.56052899e-03,
         6.10376399e-03,  7.37538149e-03,  1.01729400e-03,
         3.15361139e-02,  2.08545270e-02,  2.28891150e-03,
         4.57782299e-03,  1.27161750e-03,  1.09359105e-02,
         2.87385554e-02,  4.80671414e-02,  3.25534079e-02,
         8.77416073e-02,  4.06917599e-03,  2.54323499e-03,
         1.01729400e-03,  2.79755849e-03,  4.06917599e-03,
         3.07731434e-02,  1.27161750e-03,  3.08748728e-01,
         2.11088505e-01,  7.12105799e-03,  1.88199390e-02,
         2.28891150e-03,  2.79755849e-03,  2.54323499e-04,
         2.54323499e-03,  7.92472024e-01,  2.07527976e-01],
       [ 3.96871821e+00,  4.37716101e+01, -7.96252793e+

In [ ]:
y_pred_covid = nb_clf_co.predict(X_covid_test)
conf_mat_covid = confusion_matrix(
        y_covid_test, y_pred_covid, labels = [0, 1, 2])
conf_mat_covid

array([[679, 193, 116],
       [469, 371, 134],
       [ 50,  31, 906]])

From the variance results, three features: `Age_Group`, `Reporting_PHU_Latitude`,	`Reporting_PHU_Longitude` have variance larger than 1, the other features are all lower than 1. This is true for all three labels.

From the mean results, these three features again, with absolute value much higher than other features, and this is true for all three labels.

This result is kind of different from previous results of decision tree. Recall previously we find `Age_Group` and `Outbreak` are the two important features, but we missed `Latitude` and `Longitude`. Notice that this is the results from decision tree of depth of 3, it could be possible that we are missing some important features, especially previous seperation performance is not good enough. We then rerun the tree plot for depth=5, and features `Latitude` and `Longitude` did show up as important features. (The figure for depth=5 is too large to fit the screen thus omitted here.)

From the results of confusion matrix, we can find the case 'fatal' is better seperated compared to 'Resolved' and 'Not Resolved'. The latter two features are especially not well seperated with each other. This result is consistent with what we get from previous decision tree plot.